### PR TITLE
JDBC: Query data from ranges outside of partition range

### PR DESCRIPTION
**Pull request by leskin-in**
Modify JDBC partitioning in order to make it a perfomance-increasing feature **not** affecting the SELECT query output.

In the current version of this PR, [this](https://github.com/greenplum-db/pxf/pull/170#issuecomment-495351862), [this](https://github.com/greenplum-db/pxf/pull/170#issuecomment-495411107) and [this](https://github.com/greenplum-db/pxf/pull/170#issuecomment-498942643) comments are taken into account. They provide an explanation of the motivation for the changes.

The partition boundaries (in `INT` and `DATE` partitions) given in `RANGE` setting of `LOCATION` clause are now considered to form a closed interval (including both its lower and upper boundaries). Previously, the interval they formed was considered a left-closed one (not including its upper boundary).

#### Example
The following example illustrates the changes in API.

Consider the following string (a part of `LOCATION` clause):

`...&PARTITION_BY=column:int&RANGE=1:5&INTERVAL=2&...`

Old behaviour led to generation of fragments with constraints:
1. `>= 1 AND < 3`
2. `>= 3 AND < 5`

New behaviour leads to generation of fragments with constraints:
1. `< 1`
2. `>= 5`
3. `>= 1 AND < 3`
4. `>= 3 AND < 5`
5. `IS NULL`

#### Code
To implement the changes in API, a new (sub)package `org.greenplum.pxf.plugins.jdbc.partitioning` is introduced.

Each partition type is processed by enum `PartitionType`. The results are serialized using [Apache SerializationUtils](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/SerializationUtils.html) (this library is already used in PXF). This enables `SQLQueryBuilder` to use a common interface for all partition types and moves all partitioning logic to `org.greenplum.pxf.plugins.jdbc.partitioning` package and `JdbcPartitionFragmenter` class.

#### Other changes
* Update README.md
* Add unit tests for the new feature and update the existing ones

### CODE REVIEW

**Comment by oliverralbertini** [comment]
I noticed you had a flaky build, and we recently made an attempt to [improve this](https://github.com/greenplum-db/pxf/commit/241b29b58e29e28c969537081c0f82b8cb8671a4). To get this change you should rebase your branch onto master. Hopefully this helps get you more stable builds.

**Comment by denalex** [comment]
We had a big discussion around this recently, driven by the following 3 statements:

1. there's currently a usability issue with RANGE partition, where the right number is not included, so if users want to partition the dataset by months and specify `RANGE=1:12` they will miss data from December as 12 is excluded from the partitioning, they would need to specify `1:13` as values which is not intuitive and was actually reported as a bug.

2. current partitioning (RANGE specifically, maybe others) provide implicit data filters, which should not be the case. For example, if I say `RANGE=3:10&INTERVAL=1` -- this should be considered a HINT to the system on how to partition data, not limiting dataset to March-October window only. If the dataset needs to be limited, people should use WHERE clause statements. The idea being that if for some reason PARTITION_BY information is removed from the table DDL, then the results returned should be the same.

3. Having a lot of partitions sometimes is not desirable due to limitations on the number of connections to the external system and the overhead of running multiple table scans on the remote system. We believe that the optimal number of partitions should be no more than the number of GPDB segments, that way you can achieve the highest throughput, as if there's more than 1 partition for a segment to run, it will retrieve them sequentially, defeating the purpose.

So, in light of this, we have the following proposal, somewhat similar in spirit to this PR, but with a bit different semantics and implementation:

1. By default, make RANGE partition values INCLUSIVE on the boundaries and open from -infinity on the lower end to infinity on the upper end. In our requirement, specifying `RANGE=a:b` will be the same as specifying `RANGE=:a:b:` in the current PR proposal. This is the right way to make sure we are not restricting the data set and only provide hints to parallelize data access.
In case of partitioning by month, users will have specify `RANGE=1:12&INTERVAL=1` to include all months and get 12 partitions, or `RANGE=4:6` to include ALL months and get 5 partitions (Jan-Mar, Apr, May, June, July-Dec). If they only want to get Apr-Jun data they will also have to add `WHERE month between 4 and 6` to the query.

2. Implementation-wise, this means adding 2 more implicit partitions at Fragmenter level for `col < MIN` and `MAX <= col`, assuming `RANGE=MIN:MAX` statement.

3. Optimization to 2) above would be where users will be sure that there are no values below MIN or above MAX (like is the case with months) -- they can EXPLICITLY indicate this with the syntax `RANGE=[MIN:MAX]`, telling us not to look beyond MIN/MAX, effectively limiting the dataset returned and taking this responsibility on themselves explicitly. This will avoid adding lower partition `month < 1` which is guaranteed to bring no data and would just waste system resources. On the upper bound, we still will have to append `12 <= month`, so there is no actual implementation difference between `RANGE=[1:12]` and `RANGE=[1:12`, unless we decide to merge the upper bound in the previous partition (instead of doing `11 <= month AND month <12` along with `12 <= month` for 2 partitions we can do `11 <= month AND month <=12`, collapsing Nov and Dec data into 1 partition)

4. Additionally, we would want to introduce new partitioning option, let's call it `PARALLEL` for now. This will give the Fragmenter the hint on the optimal number of partitions to create and we will advise users to set it to equal to the total number of segments or less. Then, for example given `RANGE=[1:12]&PARALLEL=12` would achieve the same goal or specifying `RANGE=0:62000&PARALLEL=64` will result in 64 partitions each worth of 1000 step interval capturing outliers as well.

So, in summary:
- change semantics to include all data set below and above range boundaries
- potentially optimize to let users close the range with "[ ]", but still inclusive of boundary values
- introduce new PARALLEL option to better manage number of partitions explicitly

@leskin-in @frankgh @shivzone @oliverralbertini

**Comment by leskin-in** [comment]
@denalex ,
thank you for a detailed description. I completely agree with all the ideas you propose.

In fact, at Arenadata, we had exactly same idea for the feature you call `PARALLEL` about two weeks ago. We planned to implement it soon. It's a pity PXF [already knows](https://github.com/greenplum-db/pxf/blob/241b29b58e29e28c969537081c0f82b8cb8671a4/server/pxf-api/src/main/java/org/greenplum/pxf/api/model/RequestContext.java#L257) the total number of segments, but cannot automatically create partitions using that information.

Taking into consideration that no changes on GPDB side are required, I suppose this can be implemented quite soon. I could implement and include it in the current or a new PR.

##### Syntax changes
As for other changes in syntax (note the proposed ones break backward compatibility), I would propose another, more radical idea.

We could switch to other, *less wordy*, "Python-like" partitioning syntax. Examples: `PARTITION=part_column:int:1:101:10`, `PARTITION=part_column:date:2014-07:2015-07:1:month`, `PARTITION=part_column:enum:en1:en2:en3:en4:en5:en6`.

I suppose this would not only make LOCATION clause shorter, but also easier to read and understand:
* The whole partition definition would be placed a single DDL option (instead of three, which can also be reordered);
* Both [`INTERVAL`](https://en.oxforddictionaries.com/definition/interval) and [`RANGE`](https://en.oxforddictionaries.com/definition/range) mean quite similar things. It takes time to read documentation to understand which one is responsible for what.

This is an idea. At Arenadata, we discussed this a few weeks ago, and concluded this is quite radical and not of much use. Yet, as you mentioned syntax changes in your comment, I think this idea is worth thinking over.

##### Conclusion
Finally, I would like to note that all the features above can (should?) be implemented in separate PRs.

**Comment by denalex** [comment]
@leskin-in -- we just discussed you proposal, and while it makes sense in general, we felt it would be indeed a too radical change that will break backward compatibility, will be confusing to document and explain to users 2 different semantics of partitioning, granted they already have trouble understanding it.

I think we can consider our proposed solution as a bug fix, a bug being not including data outside of ranges. The proposal is in line what Spark's JDBC RDD is doing to partition the data: https://stackoverflow.com/a/42724063 by including data outside min/max.

+1 for doing separate PRs, I think we can do this in 3 steps:
1. Keep the current syntax, fix the bug by including 2 more partitions outside the ranges.
2. Potentially introduce a new additional syntax of adding "[" or "]" to NOT look outside a particular boundary. Some people here believe this is an unnecessary optimization and confuses users, so we might still revisit the need for it.
3. Potentially introduce PARALLEL option to give users more control on the number of partitions to be used (similar how Spark does it).

Please note that we do know the total number of segments (RequestContext.getTotalSegments()) so we can automatically decide the number of partitions in the Fragmenter if we like. However, this should serve as an upper bound, actual number can be smaller.

@lisakowen -- please take a look at all this discussion, what do you think will be most easily understood by users ?

**Comment by denalex** [comment]
@leskin-in -- this looks pretty good. Just minor nit-picks on documentation and recommendations on code re-organization to reduce complexity and utilize generalization. What do you think ?

Will need to spend a bit more time to review the actual new logic for processing boundaries, but seems you have good unit tests, so I hope it is all good.

The questions about ENUM and NULL handling is more serious one, though. Somehow it escaped our reasoning originally, but I think we should capture NULL values on all types and other types on ENUMs.

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/JdbcPartitionFragmenter.java [hunk 1]
```diff
@@ -91,9 +94,14 @@ public void initialize(RequestContext context) {
...
         String rangeStr = context.getOption("RANGE");
         if (rangeStr != null) {
-            range = rangeStr.split(":");
-            if (range.length == 1 && partitionType != PartitionType.ENUM) {
-                throw new IllegalArgumentException("The parameter 'RANGE' must specify ':<end_value>' for this PARTITION_TYPE");
+            if (partitionType == PartitionType.ENUM) {
+                range = rangeStr.split(":");
+            }
+            else {
+                range = rangeStr.split(":", 2);
```
[the remark this answers is not in the document]
Comment (leskin-in) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/JdbcPartitionFragmenter.java]: Yes, this will force up to two splits. Three splits syntax is not supported now; instead, [the API is changed](https://github.com/greenplum-db/pxf/pull/170#issuecomment-495411107)

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/JdbcPartitionFragmenter.java [hunk 2]
```diff
@@ -101,30 +98,10 @@ public void initialize(RequestContext context) {
                 range = rangeStr.split(":");
             }
             else {
-                range = rangeStr.split(":", 4);
-                if (range.length == 1) {
+                range = rangeStr.split(":", 2);
+                if (range.length != 2) {
                     throw new IllegalArgumentException("The parameter 'RANGE' must specify ':<end_value>' for this PARTITION_BY type");
```
[the remark this answers is not in the document]
Comment (leskin-in) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/JdbcPartitionFragmenter.java]: Thank you for your comment. I have [changed](https://github.com/greenplum-db/pxf/pull/170/files#diff-0d3419e9a6bf82afdc1af5f7d62020d2R122) this error message to `The parameter 'RANGE' has incorrect format. The correct format for this 'PARTITION_BY' type is '<start_value>:<end_value>'`

**Comment by leskin-in** [comment]
@denalex ,
Thank you for your comment. https://github.com/greenplum-db/pxf/pull/170/commits/3e358f6c838b7e66da8f0a86c534952f5614b819 addresses your remark about `NULL` and "outside-of-range" enum values.

In order to add `IS NULL` constraint, I have introduced a new partition type. We can disallow the user to pass it to `JdbcPartitionFragmenter`. I think this does not make a change, though, so no extra check was introduced.

Comments on any part of the implementation are welcome.

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/partitioning/NullPartition.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ * under the License.
+ */
+
+import org.greenplum.pxf.plugins.jdbc.utils.DbProduct;
+import org.greenplum.pxf.plugins.jdbc.partitioning.PartitionType;
+
+import java.io.Serializable;
+
+public class NullPartition implements JdbcFragmentMetadata, Serializable {
+    private static final long serialVersionUID = 1L;
```
Comment (denalex) [inline server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/partitioning/NullPartition.java]: random UID ?

File: server/pxf-jdbc/src/test/java/org/greenplum/pxf/plugins/jdbc/partitioning/EnumPartitionTest.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        EnumPartition partition = new EnumPartition(COL_RAW, new String[]{"enum1", "enum2", "enum3"});
+        String constraint = partition.toSqlConstraint(QUOTE, dbProduct);
+
+        assertEquals(
+            COL + " <> 'enum1' AND " + COL + " <> 'enum2' AND " + COL + " <> 'enum3'",
+            constraint
+        );
+    }
+
+    @Test(expected = AssertionError.class)
```
Comment (denalex) [inline server/pxf-jdbc/src/test/java/org/greenplum/pxf/plugins/jdbc/partitioning/EnumPartitionTest.java]: we are using @Rule method to check for expected exception and its message

File: server/pxf-jdbc/README.md [hunk 5]
```diff
@@ -313,35 +313,58 @@ SELECT name, count FROM dept_report WHERE max > 10000;
...
+`LOCATION` clause containing `PARTITION_BY=id:int&RANGE=1:5&INTERVAL=2` will make PXF produce 5 fragments to cover all data (real operations between column values and constants are displayed):
+1. `< 1`
+2. `>= 1 AND <= 2`
+3. `>= 3 AND <= 4`
+4. `= 5`
+5. `> 5`
+
+
+##### `ENUM` partitions
+`ENUM` partitions affect the output of the plugin: only rows with one of the `ENUM`s provided are selected.
```
[the remark this answers is not in the document]
Comment (leskin-in) [reply server/pxf-jdbc/README.md]: Fixed in https://github.com/greenplum-db/pxf/pull/170/commits/3e358f6c838b7e66da8f0a86c534952f5614b819.

For `ENUM` partitions, an [extra partition](https://github.com/greenplum-db/pxf/pull/170/commits/3e358f6c838b7e66da8f0a86c534952f5614b819#diff-d96d5dc1a93fae4a9500e7239794eac9R50) with an "exclusion" constraint [is used](https://github.com/greenplum-db/pxf/pull/170/commits/3e358f6c838b7e66da8f0a86c534952f5614b819#diff-0d3419e9a6bf82afdc1af5f7d62020d2R287) to capture values other than listed enums.

For all partitions (including `ENUM`), an [extra partition](https://github.com/greenplum-db/pxf/pull/170/commits/3e358f6c838b7e66da8f0a86c534952f5614b819#diff-4dd08e5bc8722ba2f272e10591b8e97fR27) with `IS NULL` constraint [is added](https://github.com/greenplum-db/pxf/pull/170/commits/3e358f6c838b7e66da8f0a86c534952f5614b819#diff-0d3419e9a6bf82afdc1af5f7d62020d2R303).

File: server/pxf-jdbc/README.md [hunk 6]
```diff
@@ -313,35 +313,58 @@ SELECT name, count FROM dept_report WHERE max > 10000;
...
+* **[Partition Range](#partition-range)** indicates the range of data to form partitions on. It must be in special format, depending on a type of partition:
     * If the partition type is `ENUM`, format is `<value>:<value>[:<value>[...]]`. Each `<value>` forms its own fragment;
-    * If the partition type is `INT`, format is `<start_value>:<end_value>`. PXF considers values to form a finite left-closed range (`... >= start_value AND ... < end_value`);
-    * If the partition type is `DATE`, format is `<start_date>:<end_date>`, and each date must be in format `yyyy-MM-dd`. PXF considers dates to form a finite left-closed range (`... >= start_date AND ... < end_value`);
+    * If the partition type is `INT`, format is `<start_value>:<end_value>`. PXF considers values to form a closed interval (`... >= start_value AND ... <= end_value`);
+    * If the partition type is `DATE`, format is `<start_value>:<end_value>` and each date must be in format `yyyy-MM-dd`. PXF considers dates to form a closed interval (`... >= start_value AND ... <= end_value`);
```
[the remark this answers is not in the document]
Comment (leskin-in) [reply server/pxf-jdbc/README.md]: That line [states](https://github.com/greenplum-db/pxf/pull/170/commits/bcbf397f12629e68e9435e0adc8efa499654a904#diff-d2f69cfc2b417ca813bc26aabca0b97fR329) these intervals are used to produce partitions. The current implementation treats them as closed intervals.

https://github.com/greenplum-db/pxf/pull/170/commits/3e358f6c838b7e66da8f0a86c534952f5614b819 changes the partitioning behaviour so that the structure of partitions does not affect the data selected.

To sum up, it is true that the interval of data to select is now open. However, the mentioned lines of the README describe how partitions are formed; when doing that, PXF treats the provided interval as a closed one.

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/partitioning/DatePartition.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+import org.greenplum.pxf.plugins.jdbc.utils.DbProduct;
+import org.greenplum.pxf.plugins.jdbc.partitioning.PartitionType;
+
+import java.io.Serializable;
+import java.sql.Date;
+import java.util.Calendar;
+
+public class DatePartition implements JdbcFragmentMetadata, Serializable {
+    private static final long serialVersionUID = 1L;
```
[the remark this answers is not in the document]
Comment (leskin-in) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/partitioning/DatePartition.java]: I believe the incremental `serialVersionUID` is better, as it would reflect the version of the class directly, and that is what `serialVersionUID` [is supposed to do](https://docs.oracle.com/javase/8/docs/api/java/io/Serializable.html).

[Here](https://stackoverflow.com/a/605832/10428388) the use of incremental `serialVersionUID` starting from `0` is proposed.

What do you think?

File: server/pxf-jdbc/README.md [hunk 8]
```diff
@@ -313,35 +313,58 @@ SELECT name, count FROM dept_report WHERE max > 10000;
...
 * `&PARTITION_BY=grade:enum&RANGE=excellent:good:general:bad`
 
 
 #### Mechanism
-If partitioning is enabled, a SELECT query is split into a set of small queries, each of which is called a *fragment*. All fragments are processed by separate PXF instances simultaneously. If there are more fragments than PXF instances, some instances will process more than one fragment; if only one PXF instance is available, it will process all fragments.
-
 Extra query constraints (`WHERE` expressions) are automatically added to each fragment to guarantee that every tuple of data is retrieved from the external database exactly once.
 
-Fragments are distributed randomly among PXF instances.
+Fragments are processed by separate PXF instances simultaneously (using multiple threads). If there are more fragments than PXF instances, some instances will process more than one fragment; if only one PXF instance is available, it will process all fragments.
```
[the remark this answers is not in the document]
Comment (leskin-in) [reply server/pxf-jdbc/README.md]: Fixed in https://github.com/greenplum-db/pxf/pull/170/commits/9a9b9b2359bfa840fbdf5af2d6c40eb3d5a748ed.

File: server/pxf-jdbc/README.md [hunk 9]
```diff
@@ -313,35 +313,58 @@ SELECT name, count FROM dept_report WHERE max > 10000;
...
+
+Round-robin scheduling is used to distribute fragments among *GPDB segments*. The first segment (which acquires the first fragment) is chosen pseudo-randomly (the seed is GPDB query transaction identifier). The distribution of fragments does not change dynamically (i.e. if one PXF instance has finished processing of fragments assigned to it, it will not "steal" fragments from other PXF instances).
+
+Infinite left-bounded and right-bounded intervals form exactly one extra fragment (each bounded range independently).
```
[the remark this answers is not in the document]
Comment (leskin-in) [reply server/pxf-jdbc/README.md]: Do you think https://github.com/greenplum-db/pxf/pull/170/commits/9a9b9b2359bfa840fbdf5af2d6c40eb3d5a748ed#diff-d2f69cfc2b417ca813bc26aabca0b97fR370 is better?

File: server/pxf-jdbc/README.md [hunk 10]
```diff
@@ -313,35 +311,61 @@ SELECT name, count FROM dept_report WHERE max > 10000;
...
 #### Syntax
-Three settings control partitioning feature:
+Three settings control the feature:
+
+* **[Partition By](#partition-by)** enables partitioning and indicates which column to use as a partition column (the column to which constraints are applied). Only one column can be a partition column. This setting must have format `<column>:<column_type>`, where:
+    * `<column>` is the name of the partition column;
+    * `<column_type>` is the data type of the partition column (hereinafter referred to as "partition type"). **Supported types** are `INT`, `DATE`, `ENUM` and `NULL`.
+        * `INT` and `DATE` partitions are for columns of the same types. These partitions are range-based (see below).
+        * `ENUM` is a partition for text columns. This is a value-based partition (see below).
+        * `NULL` is a special partition. It separates the whole dataset using two partitions with constraints `IS NULL` and `IS NOT NULL`.
```
[the remark this answers is not in the document]
Comment (leskin-in) [reply server/pxf-jdbc/README.md]: I think we should document all possible options.

Note that `NULL` partitions are not different from partitions of any other type in current implementation. It seems convenient and straightforward to use such implementation.

If we remove `NULL` partitions from the README, they become an undocumented feature. To prevent this, we then shall add an extra check in `JdbcPartitionFragmenter` or somewhere else to disallow their construction.

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/partitioning/DatePartition.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+    private final String column;
+    private final Date[] boundaries;
+    private final boolean isRightClosed;
+
```
[the remark this answers is not in the document]
Comment (leskin-in) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/partitioning/DatePartition.java]: `DATE` partitions are more complex than `INT` ones:
* They require [transformations in constructor](https://github.com/greenplum-db/pxf/pull/170/files#diff-74dd972fb67abcb5f727d6ce54d1b0d3R56)
* They [use](https://github.com/greenplum-db/pxf/pull/170/files#diff-74dd972fb67abcb5f727d6ce54d1b0d3R96) `DbProduct#wrapDate`

This means that in order to use generics, we should make a quite complex implementation that receives a function to process `boundaries` before constructing constraints for them.

Another option is to use `BasePartition<String>` for both INT and DATE partitions. But this does not eliminate the need to create a separate class for each of them with partition-type-specific logic.

Finally, not all partition types would actually use `boundaries` as an array, if all of them are derived from `BasePartition`.

Thus, I would rather create a base class to hold column name, but not `boundaries`.

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/partitioning/IntPartition.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+            }
+            else if (boundaries[1] == null) {
+                sb.append(columnQuoted).append(" > ").append(boundaries[0]);
+            }
+            else {
+                sb.append(columnQuoted).append(" >= ").append(boundaries[0]);
+                sb.append(" AND ");
+                sb.append(columnQuoted).append(" <= ").append(boundaries[1]);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (leskin-in) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/partitioning/IntPartition.java]: See https://github.com/greenplum-db/pxf/pull/170#discussion_r295260620.

I agree that the logic is common and we would benefit from its generalization. But to do that, I would propose to create a separate class with the only `static` member function that would generate the SQL constraint (`String`) being provided with boundaries (`String[]`) and a quoted column name.

What do you think?

File: server/pxf-jdbc/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (leskin-in) [reply server/pxf-jdbc/README.md]: After refactoring, `NULL` partitions can be easily made non-creatable by user, with proper error message and without extra special checks.

https://github.com/greenplum-db/pxf/pull/170/commits/db54c6a918e4774663d2f562fe4b3d81e600875f implements this and makes `NULL` non-creatable by a user.

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/partitioning/DatePartition.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+    /**
+     * Construct a DatePartition with given column, boundaries and right-closedness
+     * @param column
+     * @param boundaries
+     * @param isRightClosed
+     */
+    private DatePartition(String column, Calendar[] boundaries, boolean isRightClosed) {
+        // API checks
+        assert column != null;
```
[the remark this answers is not in the document]
Comment (leskin-in) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/partitioning/DatePartition.java]: Fixed in https://github.com/greenplum-db/pxf/pull/170/commits/8eb210f180c47f0a1c3b1b723e8e5bfaa7806f27 and a few preceding commits.

File: server/pxf-jdbc/src/test/java/org/greenplum/pxf/plugins/jdbc/partitioning/EnumPartitionTest.java (hunk 4 above)
Comment (leskin-in) [reply server/pxf-jdbc/src/test/java/org/greenplum/pxf/plugins/jdbc/partitioning/EnumPartitionTest.java]: Fixed in https://github.com/greenplum-db/pxf/pull/170/commits/0e5f0c81437d13486358fdf0a221e1bb23c0b560

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/JdbcPartitionFragmenter.java [hunk 14]
```diff
@@ -122,6 +135,11 @@ else if (partitionType == PartitionType.INT) {
             catch (NumberFormatException e) {
                 throw new IllegalArgumentException("The parameter 'RANGE' is invalid. Both range boundaries must be integers");
             }
+            if (rangeIntEnd < rangeIntStart) {
+                throw new IllegalArgumentException(String.format(
+                    "The parameter 'RANGE' is invalid. The <end_value> '%s' must be bigger than the <start_value> '%s'", rangeIntEnd, rangeIntStart
+                ));
+            }
```
[the remark this answers is not in the document]
Comment (leskin-in) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/JdbcPartitionFragmenter.java]: Implemented in https://github.com/greenplum-db/pxf/pull/170/commits/db54c6a918e4774663d2f562fe4b3d81e600875f.

Basically, the partitioning logic is moved to specific partition types, and `PartitionType` (renamed to `Partition`) is used to resolve specific partition types to the corresponding classes.

This greatly improves incapsulation.

Thank you for the comment, @denalex !

File: server/pxf-jdbc/README.md [hunk 15]
```diff
@@ -313,35 +313,58 @@ SELECT name, count FROM dept_report WHERE max > 10000;
...
+
+##### `INT` and `DATE` partitions
+In `INT` and `DATE` partitions, every fragment queries data from either closed or bounded interval (intervals are formed according to [Partition interval](#partition-interval) and [Partition Range](#partition-range) settings).
+
+`LOCATION` clause containing `PARTITION_BY=id:int&RANGE=1:5&INTERVAL=2` will make PXF produce 5 fragments to cover all data (real operations between column values and constants are displayed):
+1. `< 1`
+2. `>= 1 AND <= 2`
+3. `>= 3 AND <= 4`
+4. `= 5`
+5. `> 5`
```
[the remark this answers is not in the document]
Comment (leskin-in) [reply server/pxf-jdbc/README.md]: Implemented in https://github.com/greenplum-db/pxf/pull/170/commits/db54c6a918e4774663d2f562fe4b3d81e600875f

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/partitioning/DatePartition.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (leskin-in) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/partitioning/DatePartition.java]: There are lots of changes in https://github.com/greenplum-db/pxf/pull/170/commits/db54c6a918e4774663d2f562fe4b3d81e600875f aimed at improving abstraction of partition types.

Still, it seems there is not enough commonality between the classes, and the generics are of little use for the reasons described [above](https://github.com/greenplum-db/pxf/pull/170#discussion_r295260620).

File: server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/partitioning/IntPartition.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (leskin-in) [reply server/pxf-jdbc/src/main/java/org/greenplum/pxf/plugins/jdbc/partitioning/IntPartition.java]: The implementation involving a dedicated class is in https://github.com/greenplum-db/pxf/pull/170/commits/db54c6a918e4774663d2f562fe4b3d81e600875f#diff-7295bc9ef9af1c5ad489fa0c41c45500R22.

**Comment by leskin-in** [comment]
Many thanks for the [multiple improvements](https://github.com/greenplum-db/pxf/pull/170/commits/e34d15cccc4fd7ecfcf92187bc4c15f22375d4e4), @frankgh @denalex . I hope it will be possible to merge the PR soon.
